### PR TITLE
Task dl/j143 edit user functionality

### PR DESCRIPTION
**Pull request by Denidzio**
Screenshot from 2021-08-14 18-46-16

### CODE REVIEW

File: backend/src/services/user/user.service.ts [hunk 1]
```diff
@@ -26,6 +31,30 @@ class User {
     return this.#userRepository.getById(id);
   }
 
+  public async update(id: number, payload: UserEditPayload): Promise<TUser> {
+    const userWithSimilarEmail = await this.#userRepository.getByEmail(
+      payload.email,
+    );
+
+    if (userWithSimilarEmail && userWithSimilarEmail.id !== id) {
```
Comment (what1s1ove) [inline backend/src/services/user/user.service.ts]:
```suggestion
    if (userWithSimilarEmail?.id !== id) {
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]
remove the password from form and add the `bio` field
don't forget to fix import

**Comment by Denidzio** [comment]
Screenshot from 2021-08-15 10-31-59

File: backend/src/services/user/user.service.ts [hunk 2]
```diff
@@ -26,6 +31,35 @@ class User {
...
+        message: ErrorMessage.EMAIL_IS_ALREADY_TAKEN,
+      });
+    }
+
+    const updatedUser = { ...payload };
+    const hasPassword = Boolean(payload.password);
+
+    if (!hasPassword) {
+      const user = await this.getById(id);
+      updatedUser.password = user.password;
```
Comment (what1s1ove) [inline backend/src/services/user/user.service.ts]: but we can't change it in the payload for editing 🤔

File: frontend/src/common/types/user/user-form-payload.type.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { UserPayloadKey } from 'common/enums/enums';
+import { UserEditPayload } from 'jabber-shared/common/types/types';
+
+type UserFormPayload = Omit<UserEditPayload, UserPayloadKey.BIRTHDATE> & {
+  [UserPayloadKey.BIRTHDATE]: Date;
+};
```
Comment (what1s1ove) [inline frontend/src/common/types/user/user-form-payload.type.ts:6]: 🤔
```suggestion
type UserFormPayload = UserEditPayload & {
  [UserPayloadKey.BIRTHDATE]: Date;
};
```

File: frontend/src/store/user-profile/actions.ts [hunk 4]
```diff
@@ -1,21 +1,48 @@
...
+    const user = await userApi.getById(id);
 
-  return user;
-});
+    return user;
+  },
+);
 
-export { loadPodcasts, loadUser };
+const updateUser = createAsyncThunk<User, UserEditPayload, AsyncThunkConfig>(
```
Comment (what1s1ove) [inline frontend/src/store/user-profile/actions.ts]: you need to create your own reducer `configurate-user`

File: frontend/src/store/user-profile/reducer.ts [hunk 5]
```diff
@@ -36,6 +36,16 @@ const reducer = createReducer(initialState, (builder) => {
   builder.addCase(loadUser.rejected, (state) => {
     state.dataStatus = DataStatus.REJECTED;
   });
+  builder.addCase(updateUser.pending, (state) => {
```
Comment (what1s1ove) [inline frontend/src/store/user-profile/reducer.ts]: you need to create your own reducer for it

File: frontend/src/store/user-profile/common.ts [hunk 6]
```diff
@@ -1,6 +1,7 @@
 enum ActionType {
   LOAD_PODCASTS = 'user-profile/load-podcasts',
   LOAD_USER = 'user-profile/load-user',
+  UPDATE_USER = 'user-profile/update-user',
```
Comment (what1s1ove) [inline frontend/src/store/user-profile/common.ts]:
```suggestion
```

File: shared/src/common/types/user/user-edit-payload.type.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { UserPayloadKey } from '~/common/enums/enums';
+import { UserConfigurePayload } from './user-configure-payload.type';
+
+type UserEditPayload = UserConfigurePayload & {
+  [UserPayloadKey.PASSWORD]?: string;
```
Comment (what1s1ove) [inline shared/src/common/types/user/user-edit-payload.type.ts]:
```suggestion
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: frontend/src/store/configurate-user/actions.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+    const { userApi, notificationService } = extra;
+    const { auth } = getState();
+
+    const user = await userApi.update((<User>auth.user).id, payload);
+
+    notificationService.success(
+      NotificationTitle.SUCCESS,
+      NotificationMessage.USER_UPDATED,
+    );
+
```
Comment (what1s1ove) [inline frontend/src/store/configurate-user/actions.ts:22]: add redirect

File: frontend/src/store/root-reducer.ts [hunk 9]
```diff
@@ -20,6 +21,7 @@ const rootReducer = {
   userProfile,
   notification,
   toastr: toastrReducer,
+  configurateUser,
```
Comment (what1s1ove) [inline frontend/src/store/root-reducer.ts]:
```suggestion
  notification,
    configurateUser,
  toastr: toastrReducer,
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Review by what1s1ove — DISMISSED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**